### PR TITLE
feat(reporter): opossum reporter writes `.opossum` files

### PR DESCRIPTION
**Pull request by lennartclaas**
OpossumUI changed its standard input format to `.opossum` which is a zipped file containing an `input.json` and an optional `output.json` (see [OpossumUI](https://github.com/opossum-tool/OpossumUI/blob/main/USER_GUIDE.md#dot_opossum)). The Opossum reporter is adapted accordingly, i.e., it generates a `report.opossum` file containing an `input.json` with the ort result while the `output.json` is ignored.

### CODE REVIEW

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt [hunk 1]
```diff
@@ -501,15 +502,15 @@ class OpossumReporter : Reporter {
...
-                    .writeValue(gzipWriter, opossumInput.toJson())
+        ZipOutputStream(BufferedOutputStream(FileOutputStream(outputFile))).use { zipOutputStream ->
+            val jsonMapper = ObjectMapper()
+            val jsonBytes = jsonMapper.writeValueAsBytes(opossumInput.toJson())
+            val byteArrayInputStream = ByteArrayInputStream(jsonBytes)
+            BufferedInputStream(byteArrayInputStream).use { bufferedInputStream ->
+                val inputJsonEntry = ZipEntry("input.json")
+                zipOutputStream.putNextEntry(inputJsonEntry)
+                bufferedInputStream.copyTo(zipOutputStream, 4096)
+                zipOutputStream.closeEntry()
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt:513]: If you're fine with creating a temporary file, you could probably do instead (untested):

        val jsonFile = createOrtTempFile()
        JsonMapper().setSerializationInclusion(Include.NON_NULL).writeValue(jsonFile, opossumInput.toJson())
        jsonFile.packZip(outputFile)
        jsonFile.delete()

where `packZip()` comes from ORT's own `ArchiveUtils`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt [hunk 2]
```diff
@@ -501,15 +502,15 @@ class OpossumReporter : Reporter {
...
-            val gzipParameters = GzipParameters().apply {
-                compressionLevel = Deflater.BEST_COMPRESSION
-            }
-
-            GzipCompressorOutputStream(outputStream, gzipParameters).bufferedWriter().use { gzipWriter ->
-                JsonMapper()
-                    .setSerializationInclusion(Include.NON_NULL)
-                    .writeValue(gzipWriter, opossumInput.toJson())
+        ZipOutputStream(BufferedOutputStream(FileOutputStream(outputFile))).use { zipOutputStream ->
+            val jsonMapper = ObjectMapper()
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: No, it seems that JsonMapper is the appropriate approach.

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt (hunk 1 above)
Comment (lennartclaas) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: Nice and elegant suggestion. Works like a charm. Thanks.

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt [hunk 3]
```diff
@@ -19,21 +19,22 @@
 
 package org.ossreviewtoolkit.reporter.reporters.opossum
 
-import com.fasterxml.jackson.annotation.JsonInclude.Include
-import com.fasterxml.jackson.databind.json.JsonMapper
+import com.fasterxml.jackson.databind.ObjectMapper
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: I revised the commit message. I hope it is fine now.

File: reporter/src/funTest/kotlin/reporters/opossum/OpossumReporterFunTest.kt [hunk 4]
```diff
@@ -67,10 +67,14 @@ private fun TestConfiguration.generateReport(ortResult: OrtResult): String {
...
-    return GzipCompressorInputStream(outputFile)
-        .bufferedReader()
-        .use { it.readText() }
+    var reportAsString = ""
+    ZipInputStream(outputFile).use { zis ->
+        val entry = zis.nextEntry
+        if (entry?.name == "input.json") {
+            reportAsString = zis.readAllBytes().toString(Charsets.UTF_8)
+        }
+    }
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/reporters/opossum/OpossumReporterFunTest.kt:78]: Consider using a similar approach as in https://github.com/oss-review-toolkit/ort/pull/6628/files#diff-1c9eb4cfac3e546739eba0f24ba356c27589747bdfead7fbf65c2b448d25bb5bR429-R430 instead.

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: Just one (optional) nit: We prefer to start capitalized after the colon in the title, even if our linter does not catch this.

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt (hunk 1 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: > Works like a charm.

Except for the `prefix` bit, see my comment below.

File: reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt [hunk 5]
```diff
@@ -501,17 +499,10 @@ class OpossumReporter : Reporter {
...
-
-            GzipCompressorOutputStream(outputStream, gzipParameters).bufferedWriter().use { gzipWriter ->
-                JsonMapper()
-                    .setSerializationInclusion(Include.NON_NULL)
-                    .writeValue(gzipWriter, opossumInput.toJson())
-            }
-        }
+        val jsonFile = createOrtTempFile()
+        JsonMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL).writeValue(jsonFile, opossumInput.toJson())
+        jsonFile.packZip(outputFile, prefix = "input.json")
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply reporter/src/main/kotlin/reporters/opossum/OpossumReporter.kt]: As `createOrtTempFile()` does always create a filename with a random part, I had to use `createOrtTempDir()` and create the json file with the desired name afterwards.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**